File: src/main/java/com/example/dpop/auth_sms/AuthSmsService.java
package com.example.dpop.auth_sms;

import com.example.dpop.auth_sms.internal.AuthSmsSetup;
import com.example.dpop.auth_sms.internal.AuthSmsSetupRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.Instant;
import java.util.regex.Pattern;

@Service
public class AuthSmsService {

    private static final Pattern PHONE_PATTERN = Pattern.compile("^\\+?[0-9]{6,20}$");
    private static final SecureRandom RANDOM = new SecureRandom();

    private final AuthSmsSetupRepository repository;

    public AuthSmsService(AuthSmsSetupRepository repository) {
        this.repository = repository;
    }

    @Transactional
    public AuthSmsEnrollResult startEnrollment(String phoneNumber) {
        if (phoneNumber == null || phoneNumber.isBlank()) {
            throw new IllegalArgumentException("Telefonnummer ist erforderlich");
        }
        String normalized = normalizePhoneNumber(phoneNumber);
        if (!PHONE_PATTERN.matcher(normalized).matches()) {
            throw new IllegalArgumentException("Ungueltige Telefonnummer");
        }
        String tan = generateTan();
        Instant now = Instant.now();
        AuthSmsSetup setup = repository.save(new AuthSmsSetup(normalized, tan, false, now, now));
        sendTestSms(normalized, tan);
        return new AuthSmsEnrollResult(new EnrollmentRef(setup.getId()), tan);
    }

    @Transactional
    public void confirmEnrollment(EnrollmentRef ref, String tan) {
        if (ref == null || tan == null || tan.isBlank()) {
            throw new IllegalArgumentException("ref und TAN sind erforderlich");
        }
        AuthSmsSetup setup = repository.findById(ref.id())
                .orElseThrow(() -> new IllegalArgumentException("SMS-Enrollment nicht gefunden: " + ref.id()));
        if (!setup.getTan().equals(tan.trim())) {
            throw new IllegalArgumentException("Ungueltige TAN");
        }
        setup.setValidated(true);
        setup.setUpdatedAt(Instant.now());
        repository.save(setup);
    }

    @Transactional
    public AuthSmsChallengeResult startChallenge(EnrollmentRef ref) {
        AuthSmsSetup setup = repository.findById(ref.id())
                .orElseThrow(() -> new IllegalArgumentException("SMS-Enrollment nicht gefunden: " + ref.id()));
        if (!setup.isValidated()) {
            throw new IllegalArgumentException("SMS-Enrollment wurde noch nicht validiert");
        }
        String tan = generateTan();
        setup.setTan(tan);
        setup.setUpdatedAt(Instant.now());
        repository.save(setup);
        sendTestSms(setup.getPhoneNumber(), tan);
        return new AuthSmsChallengeResult(ref, tan);
    }

    @Transactional
    public void verifyChallenge(EnrollmentRef ref, String tan) {
        if (ref == null || tan == null || tan.isBlank()) {
            throw new IllegalArgumentException("ref und TAN sind erforderlich");
        }
        AuthSmsSetup setup = repository.findById(ref.id())
                .orElseThrow(() -> new IllegalArgumentException("SMS-Enrollment nicht gefunden: " + ref.id()));
        if (!setup.getTan().equals(tan.trim())) {
            throw new IllegalArgumentException("Ungueltige TAN");
        }
    }

    public boolean isValidPhoneNumber(String phoneNumber) {
        if (phoneNumber == null || phoneNumber.isBlank()) return false;
        return PHONE_PATTERN.matcher(normalizePhoneNumber(phoneNumber)).matches();
    }

    private String normalizePhoneNumber(String phoneNumber) {
        return phoneNumber.replaceAll("\\s+", "").trim();
    }

    private String generateTan() {
        int tan = RANDOM.nextInt(900000) + 100000;
        return String.valueOf(tan);
    }

    private void sendTestSms(String phoneNumber, String tan) {
        System.out.println("[MOCK SMS] TAN " + tan + " an " + phoneNumber + " versandt.");
    }
}

File: src/main/kotlin/com/example/dpop/auth_sms/AuthSmsService.kt
package com.example.dpop.auth_sms

import com.example.dpop.auth_sms.internal.AuthSmsSetup
import com.example.dpop.auth_sms.internal.AuthSmsSetupRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.security.SecureRandom
import java.time.Instant
import java.util.regex.Pattern

@Service
class AuthSmsService(private val repository: AuthSmsSetupRepository) {

    @Transactional
    fun startEnrollment(phoneNumber: String): AuthSmsEnrollResult {
        if (phoneNumber.isBlank()) {
            throw IllegalArgumentException("Telefonnummer ist erforderlich")
        }
        val normalized = normalizePhoneNumber(phoneNumber)
        if (!PHONE_PATTERN.matcher(normalized).matches()) {
            throw IllegalArgumentException("Ungueltige Telefonnummer")
        }
        val tan = generateTan()
        val now = Instant.now()
        val setup = repository.save(AuthSmsSetup(normalized, tan, false, now, now))
        sendTestSms(normalized, tan)
        return AuthSmsEnrollResult(EnrollmentRef(setup.id), tan)
    }

    @Transactional
    fun confirmEnrollment(ref: EnrollmentRef, tan: String) {
        requireNotNull(ref) { "ref und TAN sind erforderlich" }
        require(tan.isNotBlank()) { "ref und TAN sind erforderlich" }
        val setup = repository.findById(ref.id ?: throw IllegalArgumentException("Ungueltige Enrollment-Ref"))
            .orElseThrow { IllegalArgumentException("SMS-Enrollment nicht gefunden: ${ref.id}") }
        if (setup.tan != tan.trim()) {
            throw IllegalArgumentException("Ungueltige TAN")
        }
        setup.validated = true
        setup.updatedAt = Instant.now()
        repository.save(setup)
    }

    @Transactional
    fun startChallenge(ref: EnrollmentRef): AuthSmsChallengeResult {
        val setup = repository.findById(ref.id ?: throw IllegalArgumentException("Ungueltige Enrollment-Ref"))
            .orElseThrow { IllegalArgumentException("SMS-Enrollment nicht gefunden: ${ref.id}") }
        if (!setup.validated) {
            throw IllegalArgumentException("SMS-Enrollment wurde noch nicht validiert")
        }
        val tan = generateTan()
        setup.tan = tan
        setup.updatedAt = Instant.now()
        repository.save(setup)
        sendTestSms(setup.phoneNumber, tan)
        return AuthSmsChallengeResult(ref, tan)
    }

    @Transactional
    fun verifyChallenge(ref: EnrollmentRef, tan: String) {
        requireNotNull(ref) { "ref und TAN sind erforderlich" }
        require(tan.isNotBlank()) { "ref und TAN sind erforderlich" }
        val setup = repository.findById(ref.id ?: throw IllegalArgumentException("Ungueltige Enrollment-Ref"))
            .orElseThrow { IllegalArgumentException("SMS-Enrollment nicht gefunden: ${ref.id}") }
        if (setup.tan != tan.trim()) {
            throw IllegalArgumentException("Ungueltige TAN")
        }
    }

    fun isValidPhoneNumber(phoneNumber: String?): Boolean {
        if (phoneNumber.isNullOrBlank()) return false
        return PHONE_PATTERN.matcher(normalizePhoneNumber(phoneNumber)).matches()
    }

    private fun normalizePhoneNumber(phoneNumber: String): String =
        phoneNumber.replace("\\s+".toRegex(), "").trim()

    private fun generateTan(): String = (RANDOM.nextInt(900_000) + 100_000).toString()

    private fun sendTestSms(phoneNumber: String?, tan: String?) {
        println("[MOCK SMS] TAN $tan an $phoneNumber versandt.")
    }

    companion object {
        private val PHONE_PATTERN = Pattern.compile("^\\+?[0-9]{6,20}$")
        private val RANDOM = SecureRandom()
    }
}
